### PR TITLE
Simplified error handling for formspree-core. 

### PR DESCRIPTION
**Pull request by colevscode**
* Simplified error handling for formspree-core. 
* Removed return value from formspree-react submitHandler
* Included react as peer dependency of formspree-react
* Fixed dependencies and updated testing libraries

### CODE REVIEW

File: packages/formspree-react/package.json [hunk 1]
```diff
@@ -62,11 +66,14 @@
...
     "react-dom": "^18.2.0",
-    "react-test-renderer": "^17.0.2",
     "sort-package-json": "^1.22.1",
     "tsup": "^6.2.2",
     "typescript": "^4.7.4"
   },
+  "peerDependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0"
+  },
```
Comment (bhongy) [inline packages/formspree-react/package.json:76]: 👍

File: packages/formspree-core/src/utils.ts [hunk 2]
```diff
@@ -171,12 +171,15 @@ export const handleSCA = async ({
...
-export function handleLegacyErrorPayload({
-  body,
-  response
-}: SubmissionResponse): SubmissionResponse {
-  if (!hasErrors(body) && (body as any)?.error) {
-    body = { errors: [{ message: (body as any).error }] };
+export function handleLegacyErrorPayload(
+  response: SubmissionResponse
+): SubmissionResponse {
+  let body = response.body as any;
```
Comment (bhongy) [inline packages/formspree-core/src/utils.ts:177]: Do we still need `as any` here? Also, `body` is not reassigned so we can use `const`.

File: packages/formspree-react/test/useForm.test.js [hunk 3]
```diff
@@ -1,25 +1,25 @@
 import React from 'react';
 import { FormspreeProvider, useForm } from '../src';
-import ReactDOM from 'react-dom';
-import ReactTestUtils from 'react-dom/test-utils';
+import ReactDOM from 'react-dom/client';
+import { act, screen, waitFor } from '@testing-library/react';
+import user from '@testing-library/user-event';
```
Comment (bhongy) [inline packages/formspree-react/test/useForm.test.js:5]: @colevscode seems like there're a lot of React 18 + test library upgrade specifics (IIRC `@testing-library` needs upgraded for React 18). Is it possible to extract those changes to a separate PR?

File: packages/formspree-react/test/ValidationError.test.tsx [hunk 4]
```diff
@@ -19,7 +17,7 @@ afterEach(() => {
 
 it('renders a field error if one exists', () => {
   act(() => {
-    ReactDOM.render(
+    ReactDOM.createRoot(container).render(
```
Comment (bhongy) [inline packages/formspree-react/test/ValidationError.test.tsx:20]: Same question: https://github.com/formspree/formspree-js/pull/29/files?diff=split&w=1#r1211386328

File: packages/formspree-core/src/forms.ts [hunk 5]
```diff
@@ -67,22 +67,23 @@ export function isKnownError(error: FormError): error is KnownError<FormError> {
...
-  return (body as ErrorBody).errors !== undefined;
-}
-
-export interface SubmissionResponse {
-  body: SubmissionBody;
-  response: Response | null;
+export function hasErrors(
+  response: SubmissionResponse
+): response is ErrorResponse {
+  return (response as ErrorResponse).body.errors !== undefined;
```
Comment (bhongy) [inline packages/formspree-core/src/forms.ts:88]: This is fine.

Alternatively, if we use discriminated union either via a field tag, users don't have to import the type assertion function; I think it's easier to use and we don't have to have this function.

Here is one example:

```typescript
type SubmissionResult<R> = SubmissionSuccess<R> | SubmissionFailed;

type SubmissionSuccess<R> = {
  kind: 'success';
  id: string;
  data: R;
  response: Response | null;
};

type SubmissionFailed = {
  kind: 'failed';
  errors: Error[];
  response: Response | null;
};

function doSomething<R>(result: SubmissionResult<R>) {
  if (result.kind === 'success') {
    return result.data;
  }
  return result.errors;
}
```

Also, I generally try to avoid doing type assertion (`as MyType`) since it can easily hide wrong type

File: packages/formspree-core/src/forms.ts [hunk 6]
```diff
@@ -67,22 +67,23 @@ export function isKnownError(error: FormError): error is KnownError<FormError> {
...
+    data: object;
+  };
+  response: Response | null;
 }
 
-export interface ErrorBody {
-  errors: FormError[];
+export interface ErrorResponse {
+  body: { errors: FormError[] };
+  response: Response | null;
```
Comment (bhongy) [inline packages/formspree-core/src/forms.ts:80]: Can you give an example how `response` is used? Also, why is it nullable?

File: packages/formspree-core/src/forms.ts (hunk 5 above)
Comment (bhongy) [reply packages/formspree-core/src/forms.ts:88]: Thinking about this some more. I think `hasError` is the good public API so users don't rely on the specific shape of the result object.

File: packages/formspree-core/src/forms.ts [hunk 7]
```diff
@@ -67,22 +67,23 @@ export function isKnownError(error: FormError): error is KnownError<FormError> {
   );
 }
 
-export interface SuccessBody {
-  id: string;
-  data: object;
+export interface SuccessResponse {
```
[the remark this answers is not in the document]
Comment (colevscode) [reply packages/formspree-core/src/forms.ts:70]: How about SubmissionError and SubmissionSuccess?

**Comment by bhongy** [comment]
@colevscode let's close this in favor of https://github.com/formspree/formspree-js/pull/47 ?
